app/controllers/tasks_controller.rb

```diff
@@ -0,0 +1,110 @@
+require 'date'
+
+class TasksController < ApplicationController
+
+  #Controller Actions are always methods!
+  def index # index means list all
+    @tasks = Task.order(:name)
+  end
+
+  def show # details of an instance of an object
+    task_id = params[:id]
+    @task = Task.find_by(id: task_id)
+
+    if @task.nil?
+      redirect_to tasks_path # go to the index so we can see the task list
+      return
+    end
+  end
+
+  def update
+    @task = Task.find_by(id: params[:id])
+    if @task.nil?
+      redirect_to tasks_path
+      return
+    elsif @task.update(task_params)# using strong params
+
+    # elsif @task.update(
+    #   name: params[:task][:name],
+    #   description: params[:task][:description],
+    #   completed_at: params[:task][:completed_at]
+    # )
+    redirect_to tasks_path # go to the index to verify the task in the list
+      return
+    else # save failed
+      render :edit # show the task form view again
+      return
+    end
+  end
+
+  def edit
+    @task = Task.find_by(id: params[:id])
+    if @task.nil?
+      head :not_found
+      return
+    end
+  end
+
+  def unmark_complete
+    @task = Task.find_by(id: params[:id])
+    if !@task.complete?
+      redirect_to tasks_path 
+    else
+      @task.update(
+        completed_at: nil
+      )
+      redirect_to tasks_path 
+      return
+    end
+  end
+
+  def mark_complete
```

Comment:
Currently as soon as you make a task `complete` is truthy so that the view says "unmark complete" even if there's nothing filled in for completed_at. This behavior seems not quite right.

Note that `completed_at` and `complete` are redundant attributes. Consider what should your code do if `completed` is `true`, but the `completed at` is `nil`? Instead you might say that a `nil` value for `completed_at` indicates the task is incomplete, and a non-`nil` value indicates it is complete.